Feature #1 Wiki PageRevision 별 변경점 확인 기능 구현과 프로젝트 내 문제점 수정

jakdalmak opened the pull request:
## 📌 PR 제목 
Wiki 도메인 잔여 MVP 기능 구현
---
## 🤔 연관된 이슈
#7  - feature : Wiki 도메인 잔여 MVP 기능 구현


## ✨ 요약
1. PageRevision 엔티티에 대한 Create, Read 기능이 구현되었습니다.
2. 위 기능에 필요한 최소조건인 UserInfo, WikiPage 엔티티에대한 Read 기능이 구현되었습니다.
3. 위 기능 구현 외에 본 프로젝트에 산재한 각종 문제를 수정하였습니다.

---

## ✅ 주요 작업
(아래는 작업 내역에 맞춰 수정해주세요)

### 기존 작업 내역
- [x] 주된 목표인 diff 알고리즘에 대해 학습한다.
- [x] 2개 또는 N개의 PageRevision class를 DB에서 가져온다.(이하 Revision List로 명칭)
- [x] Revision List의 각 PageRevision이 지니는 String content에 대해 변경 내역을 확인한다.
- [x] 변경 내역 확인을 위한 방법을 diff 알고리즘 중 어떤 상세 알고리즘을 사용할지, 또는 다른 방법을 사용할 지 결정한다.
- [x] 결과로 산출된 변경 내역이 어떠한 양식으로 제공되는지 확인한다.
- [x] 프로토타입 기능을 설계 및 구현한다.


### 추가 작업 내역
- [x] 엔티티에 연관관계가 정상적으로 매핑되어있지 않던 문제를 수정한다.
- [x] schema.sql, data.sql에 기반한 db 환경 구현 및 더미데이터 기입 가능하게 수정한다.
- [x] PageRevision 작성을 위해 필요한 최소한의 엔티티 Read 기능 구현한다.
- [x] 실제 나무위키 내 리비전 비교처럼 diff 결과가 반환되도록 기능을 구현한다. 
- [x] 협의된 UUID 규약에 맞도록 전치사를 설정하는 클래스를 구현한다. 


---

## 🧪 테스트 방법
- [x] 테스트 코드 자체 작성 및 동작 확인(WikiDiffServiceTest.java)
![image](https://github.com/user-attachments/assets/434cca0f-7018-4fb4-b206-6ed75c5ebe72)
 * 참고 : 붉은 글씨는 deprecated 관련 이야기입니다. 테스트에 영향을 미치지 않습니다.

- [x] 자체 http 테스트(talend API)
![image](https://github.com/user-attachments/assets/ef95e47f-6d06-4aa4-b28f-74e1b44e6080)
 * 테스트 결과 이미지
 

![diff 구현](https://github.com/user-attachments/assets/e3580f1d-2ced-40c5-97bc-a92c1d250a12)
 * 좌측 : 실제 나무위키 페이지
 * 우측 : 해당 나무위키 리비전 별 내용에 기반한 본 서비스에서의 diff 결과 html


---

## 🔍 참고 사항 및 리뷰 요구사항
(아래는 작업 내역에 맞춰 수정해주세요)

### 참고사항
주요내역 0. ★★★★  중 요  ★★★★
'ERD Cloud -> SQL기반 프로젝트 내 이전' 작업 도중 실수가 있었음에 사과드립니다.
엔티티에 연관관계 매핑이 되지않던 문제는 ERD Cloud에서 sql문 추출 시 '비식별 FK' 체크 표기를 하지 않아 발생한 문제입니다.
즉, 이전 schema.sql을 기반으로 로컬 DB를 구현하셨다면 해당 DB에는 연관관계 자체가 존재하지않습니다.
**_로컬DB 내역을 삭제 후, 본 브랜치에 포함된 schema.sql을 기반으로 다시 구현해주시기 바랍니다._**

 * 주요내역 1. 
PageRevisionCommandController에서 참조중인
UserInfoQueryService는 spring security 기반의 context-holder 통한 사용자 정보 불러오기 구현 후 삭제해야할 대상입니다.
아직 springSecurity 기반의 사용자 인증 체계가 구현되지 않아 http 단에서의 테스트(talend API 이용한 HTTP 프로토콜상에서 테스트) 시 userId를 requestParam으로 임시로 전달하고, 이 값을 임시로 가져와 사용하기 위한 목적입니다.

* 주요내역 2.
swagger 기반의 API docs가 아직 작성되지 않았습니다.
현재 프론트 측에서 에디터 관련 페이지를 다음 주에 전달해주신다고 말씀해주신만큼
백엔드 개발의 빠른 진행을 위해 우선적으로 pull request 진행하고, 추후 타 기능 구현을 위한 브랜치에서 추가 작업하겠습니다.

* 주요내역 3.
이제 spring.sql.init.mode가 always로 설정되어있을 때 이미 테이블 및 제약설정이 존재하더라도 
schema.sql과 data.sql의 구동에 이상이 생기지 않습니다.
더미데이터가 필요한 경우 양식에 맞추어 data.sql에 INSERT INTO를 수행해주세요.

* 주요내역 4.
엔티티에 대해 아래와 같이 기술하여 prefix가 포함된 UUID를 구현할 수 있도록 설정하였습니다.
```
 @PrePersist
    public void prePersist() {
        if (this.revisionId == null) {
            this.revisionId = IdPrefixUtil.get(this.getClass().getSimpleName()) + "_" + UUID.randomUUID();
        }
    }
```
이를 위해 필요로 하는 yml 파일은 microsoft teams 폴더 내 ID-PREFIX.yml 파일을 참조해주세요.


### 리뷰 요구사항
1. @Transactional 설정을 상세히 어떻게 할 지 협의한 바 없어 create 부문에만 설정해두었습니다.
2. 문법, 명명규칙(클래스명, 변수 등), 패키지 할당, 리소스 낭비 등 뭐든간에 지적해주시면 환영입니다.

jakdalmak committed:
feature: 더미 목적의 POST /write-dummydata 구현과 추가설정
더미데이터 구현과 복기 목적의 TestController 구현.
TestController는 "/api/v1"과 같이 버전이 관리됩니다.
PageRevision에 대한 기능 구현 위한 PageRevision Service, PageRevisionRepository 구현
PageRevision Entity가 created_at을 자동으로 저장하게 하기 위한 @CreatedDate 어노테이션 추가
@CreatedDate 의 정상 동작 위해 LpickBackendApplication.java에 @EnableJpaAuditing 어노테이션 추가
Spring Security의 인증 요청과 csrf 우회 위한 SecurityConfig 클래스 구현
PageRevision 관련 동작을 수행하기 위한 DTO인 Request, Response 추가

jakdalmak committed:
feat : wikiDiffService 1차 구현
문장, 단어 단위 diff 구현 성공.
나무위키 형식의 diff에는 아직 근접하지 못함.
문제점 작성 -
1. 이전버전 문장 - 이후버전 문장 과 같이 변경된 문장끼리 페어를 이루도록 json이 표기되어야함.
2.  나무위키와 다르게 문자 단위 diff 추적이 아님. 문자 단위에 맞게 json 또한 추가 변경해야함

jakdalmak committed:
chore : 알고리즘 방향 전환성 이전 임시저장
difflib:diff-match-patch과 같이 이미 어느정도 구현된 외부 라이브러리를 기반으로 구현 후
해당 라이브러리를 뜯어본 뒤 적용하는 방식으로 기능 구현 방향성 전환.
이를 위해 기존 진척도를 임시저장하는 목적으로 커밋.

jakdalmak committed:
feat : diff-match-patch 기반의 리비전 비교 기능 프로토타입 구현
PageRevisionService로 PageRevision 엔티티에 대한 기초 접근 가능
WikiDiffServiceV2에 diff-match-patch 기반, openNamu/route/view-diff.py에 기반한 기능 이관 완료

jakdalmak committed:
chore : 기능 개발을 위한 더미 클래스 및 DTO 삭제, 재명명
CQRS 패턴 채용에따른 {{엔티티명}} + Command/Query + Controller/Service/Repository 형식으로 클래스 명칭변경
사용하지 않기로 결정된 클래스 전부 삭제

jakdalmak committed:
feat : WikiDiffService에 대한 Test 코드 추가
WikiDiffService를 이용해 발생하는 예외 상황과 값 반환 양식 확인을 위한 test 코드 작성

jakdalmak committed:
DB, refactor : Entity 연관관계 수정 및 관련 기능 리팩토링
schema.sql에 비식별 fk sql이 존재하지 않던 문제 해소.
엔티티에 many to one, one to one 연관관계 구현.(one to many는 개발 도중 필요 시 추가)
본 브랜치의 pageRevision 관련 기능 상에서 연관관계 대상의 id를 string으로 기입하던 기존 방식에서 entity로 할당하도록 리팩토링

jakdalmak committed:
DB : 더미데이터 추가 가능한 data.sql 추가 및 기반 환경 구현
application.yml의 spring.sql.init.mode를 always로 고정.
ddl auto none으로 설정 유지.
schema.sql에 CREATE TABLE IF NOT EXSITS... 과 ALTER TABLE ... DROP CONSTRAINT IF EXISTS .. 적용
data.sql에 INSERT INTO ... VALUES ... ON CONFLICT ({{pk명}}) DO NOTHING 적용
위 구문들 적용함으로서 spring.sql.init.mode가 always여도, 프로젝트 재 실행시 테이블 및 연관관계 중복 생성문제로 인한 실행 취소 현상 발생하지않음.

jakdalmak committed:
fix : create 동작에 대해 transactional 기입 안되어있던 문제 수정

jakdalmak committed:
fix : transactional 무분별하게 기입된 문제 수정
create, update, delete 구문에만 우선적으로 기입 예정.
trasnsactional 전략 관련하여 논의 후 클래스 전역 transactional 설정 수행

jakdalmak committed:
feat : 기본키 UUID 규약 Prefix 작성 목적의 IdPrefixUtil 구현
엔티티명:Prefix 키밸류 쌍 형식의 ID-PREFIX.yml 파일 구현.
위 파일에 기반하여 Prefix를 가져오기위해 Static으로 Map을 저장 및 업데이트하는 IdPrefixUtil 클래스 구현.
IdPrefixUtil을 이용하여 IdPrefixUtil.get(this.getClass().getSimpleName()) + "_" + UUID.randomUUID() 와 같이 기본키 작성 가능.
ID-PREFIX.yml은 gitignore 대상이므로 디스코드를 참조해주세요.

JungUiJin commented on lpick-backend/src/main/java/com/notfound/lpickbackend/Wiki/Command/Application/Service/PageRevisionCommandService.java:
```diff
@@ -0,0 +1,59 @@
+        WikiPage targetWikiPage = wikiPageQueryService.getWikiPageById(request.getWikiId());
+
+        // 기존 버전 + 1 하기 위한 리비전 개수 카운팅
+        // count vs 제일 높은값 1개 뽑기 중 하나 sql문 효율성 비교 필요
```
개인적인 생각으로는 Wiki Id로 조회된 리비젼 중 가장 최근 리비젼의 버젼 넘버를 가져와서 +1 하는게 제일 효율적일 것 같습니다.

JungUiJin commented on lpick-backend/src/main/java/com/notfound/lpickbackend/Wiki/Command/Application/Controller/PageRevisionCommandController.java:
```diff
@@ -0,0 +1,42 @@
+     * @return PageRevisionResponse를 반환한다.
+     * */
+    // post임에도 requestParam이 쓰인이유는, SpringSecurity 기반 적용 되지 않았기 때문.
+    @PostMapping("/page-revision")
```
DB에 더미데이터 PageRevision을 만들기위한 기능이라고 적어주셨는데, 앞으로 POST API로 계속 사용하실 예정일까요? 계속 사용하실 예정이면 메소드명을 변경하는게 좋을것같습니당

JungUiJin commented on lpick-backend/src/main/java/com/notfound/lpickbackend/Wiki/Command/Application/Service/PageRevisionCommandService.java:
```diff
@@ -0,0 +1,59 @@
+    private final PageRevisionCommandRepository pageRevisionCommandRepository;
+
+    @Transactional
+    public PageRevisionResponse writeNewRevision(PageRevisionRequest request, UserInfo user) {
```
여기도 계속해서 사용할 예정이라면 write를 create로 통일하는건 어떨까욥

JungUiJin commented on lpick-backend/src/main/java/com/notfound/lpickbackend/Wiki/Query/Service/PageRevisionQueryService.java:
```diff
@@ -0,0 +1,27 @@
+public class PageRevisionQueryService {
+    private final PageRevisionQueryRepository pageRevisionQueryRepository;
+
+    public List<PageRevision> getTwoRevision(String wikiId, String oldVersion, String newVersion) {
```
여기는 DB에서 revision을 조회하는거라 read로 작성하는게 좋을것같습니닷

JungUiJin reviewed (approved):
전반적인 로직은 다 너무 잘 작성해주신것 같습니다. 주석도 자세히 달아주셔서 코드읽기가 편했네욥. 고생 많으셨어욥.

jakdalmak committed:
fix : 코드리뷰 기반 메소드 명칭 수정
메소드 명명규칙(CRUD 또는 비즈니스로직임이 표기되어야함)에 맞지않게 임시로 표기해두었던 메소드 명칭 수정
revision_number 관련사항은 간단히 논의 후 별도 브랜치에서 리팩토링 예정
